src/solver/lu_decomposition.cu

```diff
@@ -4,190 +4,199 @@
-              U[U_idx] = A[A_idx];
+              size_t U_idx = d_uik_nkj[uik_nkj_offset].first + tid;
+              size_t A_idx = d_aik[aik_offset++] + tid;
+              d_U[U_idx] = d_A[A_idx];
             }
 
-            for (size_t ikj = 0; ikj < uik_nkj[uik_nkj_offset].second; ++ikj)
+            for (size_t ikj = 0; ikj < d_uik_nkj[uik_nkj_offset].second; ++ikj)
             {
-              size_t U_idx_1 = uik_nkj[uik_nkj_offset].first + tid;
-              size_t L_idx = lij_ujk[lij_ujk_offset].first + tid;
-              size_t U_idx_2 = lij_ujk[lij_ujk_offset].second + tid;
-              U[U_idx_1] -= L[L_idx] * U[U_idx_2];
+              size_t U_idx_1 = d_uik_nkj[uik_nkj_offset].first + tid;
+              size_t L_idx = d_lij_ujk[lij_ujk_offset].first + tid;
+              size_t U_idx_2 = d_lij_ujk[lij_ujk_offset].second + tid;
+              d_U[U_idx_1] -= d_L[L_idx] * d_U[U_idx_2];
               ++lij_ujk_offset;
             }
             ++uik_nkj_offset;
           }
           // lower triangular matrix
 
-          L[lki_nkj[lki_nkj_offset++].first + tid] = 1.0;
+          d_L[d_lki_nkj[lki_nkj_offset++].first + tid] = 1.0;
 
           for (size_t iL = 0; iL < inLU.first; ++iL)
           {
-            if (device->do_aki_[do_aki_offset++])
+            if (d_do_aki[do_aki_offset++])
             {
-              size_t L_idx = lki_nkj[lki_nkj_offset].first + tid;
-              size_t A_idx = device->aki_[aki_offset++] + tid;
-              L[L_idx] = A[A_idx];
+              size_t L_idx = d_lki_nkj[lki_nkj_offset].first + tid;
+              size_t A_idx = d_aki[aki_offset++] + tid;
+              d_L[L_idx] = d_A[A_idx];
             }
-            for (size_t ikj = 0; ikj < lki_nkj[lki_nkj_offset].second; ++ikj)
+            for (size_t ikj = 0; ikj < d_lki_nkj[lki_nkj_offset].second; ++ikj)
             {
-              size_t L_idx_1 = lki_nkj[lki_nkj_offset].first + tid;
-              size_t L_idx_2 = lkj_uji[lkj_uji_offset].first + tid;
-              size_t U_idx = lkj_uji[lkj_uji_offset].second + tid;
-              L[L_idx_1] -= L[L_idx_2] * U[U_idx];
+              size_t L_idx_1 = d_lki_nkj[lki_nkj_offset].first + tid;
+              size_t L_idx_2 = d_lkj_uji[lkj_uji_offset].first + tid;
+              size_t U_idx = d_lkj_uji[lkj_uji_offset].second + tid;
+              d_L[L_idx_1] -= d_L[L_idx_2] * d_U[U_idx];
               ++lkj_uji_offset;
             }
-            L[lki_nkj[lki_nkj_offset].first + tid] /= U[device->uii_[uii_offset] + tid];
+            d_L[d_lki_nkj[lki_nkj_offset].first + tid] /= d_U[d_uii[uii_offset] + tid];
             ++lki_nkj_offset;
             ++uii_offset;
           }
         }
       }
-    }  // end of kernel
+    }  // end of CUDA kernel
+
+    /// This is the function that will copy the constant data
+    ///   members of class "CudaLuDecomposition" to the device
+    void CopyConstData(LuDecomposeConst* hostptr, LuDecomposeConst*& devptr)
```

Comment:
```suggestion
    void CopyConstData(LuDecomposeConst* hostptr, LuDecomposeConst* devptr)
```

Only the `*` is needed here

Reply:
Actually no. Because I will allocate the memory space for `devptr` inside the `CopyConstData` function, I have to add the `&` here to make sure that it is reflected outside this functions. Otherwise I will get a segmentation fault when calling the CUDA kernel. See my comments at https://github.com/NCAR/micm/blob/update_cuda_class/include/micm/solver/cuda_lu_decomposition.cuh#L16.

Reply:
I think you probably want to make this a return value of the function, and I'm not sure if you really need to allocate memory for this. I would suggest something like:
```
LuDecomponseConst CopyConstData(LuDecomposeConst& hostptr)
```
and then just have the `CudaLuDecomposition` data member be:
```
LuDecompositionConst devptr_;
```
(instead of being a pointer)

Reply:
Thanks @mattldawson . I am actually struggling with pointers in C++ a lot and I try to avoid using it unless necessary. I think what you suggest makes sense and I am working on it now.

Reply:
@mattldawson @K20shores I just updated the `CopyConstData` to return the `LuDecomponseConst` type. I still generate a local `LuDecomponseConst` struct inside the `CopyConstData` function but let me know if this is not what I am supposed to do.

Reply:
Kyle: C/C++ prefers to return the value rather than passing the argument as value/pointer.